src/server/frontend_wayland/wl_surface.cpp

```diff
@@ -102,8 +102,19 @@ mf::WlSurface::WlSurface(
 
 mf::WlSurface::~WlSurface()
 {
-    role->surface_destroyed();
-    session->destroy_buffer_stream(stream);
+    // We can't use a function try block as we want to access `client`:
+    // "Before any catch clauses of a function-try-block on a destructor are entered,
+    // all bases and non-variant members have already been destroyed."
+    try
+    {
+        // Destroy the buffer stream first, as surface_destroyed() may throw
+        session->destroy_buffer_stream(stream);
+        role->surface_destroyed();
+    }
+    catch (...)
+    {
+        mw::internal_error_processing_request(client, "WlSurface::~WlSurface()");
```

Comment:
I think this (and the equivalent in William's PR) is wrong. This isn't (necessarily) a Mir implementation error - the client could be destroying surface roles in incorrect order, which is a client protocol violation not a Mir error.

If we ensure that `surface_destroyed` is throwing an appropriate subclass of `ProtocolException` we can just add a catch handler for that before this catch-all.

Reply:
I agree this is a client protocol violation, but to throw a `ProtocolError` we need a suitable error code. `wl_surface` has "invalid_scale" and "invalid_transform" neither of which seems appropriate.

Reply:
On consideration, I don't think there's much point sending a protocol error to a deleted object. This is likely the least bad way to notify the problem.